### ISSUE TITLE
compilepkg tries to call libdefault_nogo.so

### ISSUE THREAD

**Issue by linzhp** [bazelbuild/rules_go#2470]
On our Jenkins' builds, rules_go tries to call `builder compilepkg` with libdefault_nogo.so, instead of a nogo exectuable for `@org_golang_x_crypto`, resulting errors like this:
```
ERROR: /mnt/jenkins/.cache/bazel/_bazel_jenkins/e1d11652acf0c5d23aa28905edcc7be3/external/org_golang_x_crypto/poly1305/BUILD.bazel:3:1: GoCompilePkg external/org_golang_x_crypto/poly1305/go_default_library.a failed (Exit 1): builder failed: error executing command 
  (cd /mnt/jenkins/.cache/bazel/_bazel_jenkins/e1d11652acf0c5d23aa28905edcc7be3/sandbox/linux-sandbox/7947/execroot/__main__ && \
  exec env - \
    CGO_ENABLED=1 \
    GOARCH=amd64 \
    GOOS=linux \
    GOPATH='' \
    GOROOT=bazel-out/k8-fastbuild-ST-9c9985fc97d8f8bef19f5d79323610d073a386b599d7ca1532565183d6fe3142/bin/external/io_bazel_rules_go/stdlib_ \
    GOROOT_FINAL=GOROOT \
    PATH=/usr/bin:/bin \
  bazel-out/host/bin/external/go_sdk/builder compilepkg -sdk external/go_sdk -installsuffix linux_amd64 -src external/org_golang_x_crypto/poly1305/bits_compat.go -src external/org_golang_x_crypto/poly1305/bits_go1.13.go -src external/org_golang_x_crypto/poly1305/mac_noasm.go -src external/org_golang_x_crypto/poly1305/poly1305.go -src external/org_golang_x_crypto/poly1305/sum_amd64.go -src external/org_golang_x_crypto/poly1305/sum_generic.go -src external/org_golang_x_crypto/poly1305/sum_noasm.go -src external/org_golang_x_crypto/poly1305/sum_ppc64le.go -src external/org_golang_x_crypto/poly1305/sum_s390x.go -src external/org_golang_x_crypto/poly1305/sum_amd64.s -src external/org_golang_x_crypto/poly1305/sum_ppc64le.s -src external/org_golang_x_crypto/poly1305/sum_s390x.s -src external/org_golang_x_crypto/poly1305/sum_vmsl_s390x.s -importpath golang.org/x/crypto/poly1305 -p golang.org/x/crypto/poly1305 -package_list bazel-out/host/bin/external/go_sdk/packages.txt -o bazel-out/k8-fastbuild-ST-9c9985fc97d8f8bef19f5d79323610d073a386b599d7ca1532565183d6fe3142/bin/external/org_golang_x_crypto/poly1305/go_default_library.a -nogo bazel-out/k8-opt-exec-2B5CBBC6-ST-9c9985fc97d8f8bef19f5d79323610d073a386b599d7ca1532565183d6fe3142/bin/libdefault_nogo.so -x bazel-out/k8-fastbuild-ST-9c9985fc97d8f8bef19f5d79323610d073a386b599d7ca1532565183d6fe3142/bin/external/org_golang_x_crypto/poly1305/go_default_library.x -gcflags -shared -asmflags -shared)
Execution platform: @local_config_platform//:host

Use --sandbox_debug to see verbose messages from the sandbox 
compilepkg: nogo command 'bazel-out/k8-opt-exec-2B5CBBC6-ST-9c9985fc97d8f8bef19f5d79323610d073a386b599d7ca1532565183d6fe3142/bin/libdefault_nogo.so -p golang.org/x/crypto/poly1305 -importcfg /mnt/jenkins/.cache/bazel/_bazel_jenkins/e1d11652acf0c5d23aa28905edcc7be3/sandbox/linux-sandbox/7947/execroot/__main__/bazel-out/k8-fastbuild-ST-9c9985fc97d8f8bef19f5d79323610d073a386b599d7ca1532565183d6fe3142/bin/external/org_golang_x_crypto/poly1305/importcfg719768151 -x bazel-out/k8-fastbuild-ST-9c9985fc97d8f8bef19f5d79323610d073a386b599d7ca1532565183d6fe3142/bin/external/org_golang_x_crypto/poly1305/go_default_library.x /mnt/jenkins/.cache/bazel/_bazel_jenkins/e1d11652acf0c5d23aa28905edcc7be3/sandbox/linux-sandbox/7947/execroot/__main__/external/org_golang_x_crypto/poly1305/bits_go1.13.go /mnt/jenkins/.cache/bazel/_bazel_jenkins/e1d11652acf0c5d23aa28905edcc7be3/sandbox/linux-sandbox/7947/execroot/__main__/external/org_golang_x_crypto/poly1305/poly1305.go /mnt/jenkins/.cache/bazel/_bazel_jenkins/e1d11652acf0c5d23aa28905edcc7be3/sandbox/linux-sandbox/7947/execroot/__main__/external/org_golang_x_crypto/poly1305/sum_amd64.go /mnt/jenkins/.cache/bazel/_bazel_jenkins/e1d11652acf0c5d23aa28905edcc7be3/sandbox/linux-sandbox/7947/execroot/__main__/external/org_golang_x_crypto/poly1305/sum_generic.go' exited unexpectedly: signal: segmentation fault (core dumped)
[29,202 / 29,322] GoPath src/code.uber.internal/infra/bmo/plugins/crane/firewall_mocks_gomock_gopath; 64s remote-cache, linux-sandbox ... (6 actions running)
[29,202 / 29,322] GoPath src/code.uber.internal/infra/bmo/plugins/crane/firewall_mocks_gomock_gopath; 78s remote-cache, linux-sandbox ... (6 actions running)
[29,203 / 29,322] GoStdlib external/io_bazel_rules_go/stdlib_/pkg; 66s remote-cache, linux-sandbox ... (4 actions running)
[29,203 / 29,322] GoStdlib external/io_bazel_rules_go/stdlib_/pkg; 90s remote-cache, linux-sandbox ... (4 actions running)
[29,203 / 29,322] GoStdlib external/io_bazel_rules_go/stdlib_/pkg; 114s remote-cache, linux-sandbox ... (4 actions running)
INFO: Elapsed time: 190.696s, Critical Path: 121.06s
INFO: 5957 processes: 96 remote cache hit, 5861 linux-sandbox.
FAILED: Build did NOT complete successfully
```
I am not able to reproduce the error by manually building `@org_golang_x_crypto//poly1305:go_default_library` on either Linux or Darwin. The command looks like this in manual builds:
```
(cd /home/zplin/.cache/bazel/_bazel_zplin/651ef62415c6e3a55d6e5c7b32e440dc/execroot/__main__ && \
  exec env - \
    CGO_ENABLED=1 \
    GOARCH=amd64 \
    GOOS=linux \
    GOPATH='' \
    GOROOT=external/go_sdk \
    GOROOT_FINAL=GOROOT \
    PATH=/usr/bin:/usr/lib/ccache:/bin \
  bazel-out/host/bin/external/go_sdk/builder compilepkg -sdk external/go_sdk -installsuffix linux_amd64 -src external/org_golang_x_crypto/poly1305/bits_compat.go -src external/org_golang_x_crypto/poly1305/bits_go1.13.go -src external/org_golang_x_crypto/poly1305/mac_noasm.go -src external/org_golang_x_crypto/poly1305/poly1305.go -src external/org_golang_x_crypto/poly1305/sum_amd64.go -src external/org_golang_x_crypto/poly1305/sum_generic.go -src external/org_golang_x_crypto/poly1305/sum_noasm.go -src external/org_golang_x_crypto/poly1305/sum_ppc64le.go -src external/org_golang_x_crypto/poly1305/sum_s390x.go -src external/org_golang_x_crypto/poly1305/sum_amd64.s -src external/org_golang_x_crypto/poly1305/sum_ppc64le.s -src external/org_golang_x_crypto/poly1305/sum_s390x.s -src external/org_golang_x_crypto/poly1305/sum_vmsl_s390x.s -importpath golang.org/x/crypto/poly1305 -p golang.org/x/crypto/poly1305 -package_list bazel-out/host/bin/external/go_sdk/packages.txt -o bazel-out/k8-fastbuild/bin/external/org_golang_x_crypto/poly1305/go_default_library.a -nogo bazel-out/k8-opt-exec-2B5CBBC6/bin/default_nogo_/default_nogo -x bazel-out/k8-fastbuild/bin/external/org_golang_x_crypto/poly1305/go_default_library.x -gcflags '' -asmflags '')
```

@jayconrod Do you have any idea why nogo is built into a shared library in some environment?

### What version of rules_go are you using?

fbbbfde2dff5072fe118b369a699d456ec756b0c

### What version of gazelle are you using?

2078a44d2bba287db98b2b48b5f6cf4f08642489

### What version of Bazel are you using?

3.0.0

### Does this issue reproduce with the latest releases of all the above?

It should

### What operating system and processor architecture are you using?

Linux amd64

**Comment by jayconrod** [comment]
Strange. Do you have a `go_binary` with `linkmode = "shared"` or something like that? I suspect that part of the configuration is leaking to `nogo`, which should be built for the execution configuration.

**Comment by linzhp** [comment]
Yes! You are right. I am finally able to reproduce by doing a global search on our code base for "linkmode". It turned out the `go_binary` using `@org_golang_x_crypto` has `linkmode = "c-shared"`. Let me know if you need me to prepare a minimal example to reproduce it.

**Comment by jayconrod** [comment]
I got it to reproduce. I'll try to track it down tomorrow.

I've been trying to fix another nogo issue for v0.23 (#2374). I keep finding that configuration transitions don't quite work the way I expect them to.
